MixedDenseSparseGradientModule

gcramer23 opened the pull request:
work for issue https://github.com/pytorch/pytorch/issues/54445

- Adds a MixedDenseSparseGradientModule 
- refactored SparseGradientModule tests
- added MixedDenseSparseGradientModule tests
- testing commands
  1. pytest test/distributed/test_c10d.py -k test_sparse_gradients -vs
  2.  pytest test/distributed/test_c10d.py -k test_mixed_dense_sparse_gradients -vs
  3.  pytest test/distributed/test_c10d.py -k test_ddp_comm_hook_sparse_gradients -vs
  4.  pytest test/distributed/test_c10d.py -k test_ddp_comm_hook_mixed_dense_sparse_gradients -vs

Differential Revision: [D27465517](https://our.internmc.facebook.com/intern/diff/D27465517)

gcramer23 committed:
MixedDenseSparseModule

[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"




[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

- Adds a MixedDenseSparseGradientModule sparse embedding layer to a fully connected dense layer.
- refactored existing SparseGradientModule tests to pass Model as a parameter
- added MixedDenseSparseGradientModule tests




[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

- Adds a MixedDenseSparseGradientModule sparse embedding layer to a fully connected dense layer.
- refactored existing SparseGradientModule tests to pass Model as a parameter
- added MixedDenseSparseGradientModule tests




[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

- Adds a MixedDenseSparseGradientModule sparse embedding layer to a fully connected dense layer.
- refactored existing SparseGradientModule tests to pass Model as a parameter
- added MixedDenseSparseGradientModule tests




[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

- Adds a MixedDenseSparseGradientModule sparse embedding layer to a fully connected dense layer.
- refactored existing SparseGradientModule tests to pass Model as a parameter
- added MixedDenseSparseGradientModule tests




[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

- Adds a MixedDenseSparseGradientModule sparse embedding layer to a fully connected dense layer.
- refactored existing SparseGradientModule tests to pass Model as a parameter
- added MixedDenseSparseGradientModule tests




[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

- Adds a MixedDenseSparseGradientModule sparse embedding layer to a fully connected dense layer.
- refactored existing SparseGradientModule tests to pass Model as a parameter
- added MixedDenseSparseGradientModule tests




[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

- Adds a MixedDenseSparseGradientModule 
-- fully connected dense layer to sparse embedding layer
- refactored existing SparseGradientModule tests to pass Model as a parameter
- added MixedDenseSparseGradientModule tests

tested with the following commands

SparseGradientModule
- pytest test/distributed/test_c10d.py -k test_ddp_comm_hook_sparse_gradients -vs
- pytest test/distributed/test_c10d.py -k test_sparse_gradients -vs

MixedDenseSparseGradientModule
- pytest test/distributed/test_c10d.py -k test_ddp_comm_hook_mixed_dense_sparse_gradients -vs
- pytest test/distributed/test_c10d.py -k test_mixed_dense_sparse_gradients -vs


[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

work for issue https://github.com/pytorch/pytorch/issues/54445

- Adds a MixedDenseSparseGradientModule 
-- fully connected dense layer to sparse embedding layer
- added MixedDenseSparseGradientModule tests

MixedDenseSparseGradientModule test commands
- pytest test/distributed/test_c10d.py -k test_mixed_dense_sparse_gradients -vs


[ghstack-poisoned]

mrshenli commented on test/distributed/test_c10d.py:
```diff
@@ -2223,6 +2223,17 @@ def __init__(self):
 
+class MixedDenseSparseGradientModule(nn.Module):
+    def __init__(self):
+        super(MixedDenseSparseGradientModule, self).__init__()
```
nit: this is prior to this PR. We are on Python 3.6+ now, this can be simplified to `super().__init__()`. Please feel free to edit existing tests to use PY3 conventions if you like.

gcramer23 replied:
I updated my Module.

mrshenli commented on test/distributed/test_c10d.py:
```diff
@@ -3499,6 +3510,69 @@ def test_sparse_gradients(self):
+    def test_mixed_dense_sparse_gradients_grad_is_view(self):
+        self._test_mixed_dense_sparse_gradients(gradient_as_bucket_view=True)
+
+    # @requires_gloo()
```
this is debugging vestige?

mrshenli commented on test/distributed/test_c10d.py:
```diff
@@ -3499,6 +3510,69 @@ def test_sparse_gradients(self):
+
+        # "get_future" API does not support gloo backend, see GH Issue #42048.
+        # Instead, we wait for an allreduce work, and write its result to a Future.
+        def allreduce_hook_gloo(
```
curious, why do we need a hook here? Does DDP's default comm hit any error?

gcramer23 replied:
resolved from based on the dense-sparse conversation.

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

work for issue https://github.com/pytorch/pytorch/issues/54445

- Adds a MixedDenseSparseGradientModule 
-- fully connected dense layer to sparse embedding layer
- added MixedDenseSparseGradientModule tests

MixedDenseSparseGradientModule test commands
- pytest test/distributed/test_c10d.py -k test_mixed_dense_sparse_gradients -vs


[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

work for issue https://github.com/pytorch/pytorch/issues/54445

- Adds a MixedDenseSparseGradientModule 
-- fully connected dense layer to sparse embedding layer
- added MixedDenseSparseGradientModule tests

MixedDenseSparseGradientModule test commands
- pytest test/distributed/test_c10d.py -k test_mixed_dense_sparse_gradients -vs


[ghstack-poisoned]

mrshenli commented on test/distributed/test_c10d.py:
```diff
@@ -2223,6 +2223,17 @@ def __init__(self):
+
+    def forward(self, x):
+        x = self.relu(self.fc(x.to(torch.float)))
+        return F.softmax(self.embedding(x.to(torch.long)), dim=1)
```
Can you try putting the linear layer after embedding table. It will also make more sense logically, i.e., first look up embeddings from the table and then use those embeddings to run fwd and bwd. The following code works for me locally:


```python
class SparseGradientModule(nn.Module):
    def __init__(self):
        super(SparseGradientModule, self).__init__()
        self.embedding = nn.EmbeddingBag(10, 10, sparse=True)
        self.linear = nn.Linear(10, 10)

    def forward(self, x):
        return F.softmax(self.linear(self.embedding(x)), dim=1)
```

gcramer23 replied:
updated the MixedDenseSparseModule

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

work for issue https://github.com/pytorch/pytorch/issues/54445

- Adds a MixedDenseSparseGradientModule 
- refactored SparseGradientModule tests
- added MixedDenseSparseGradientModule tests
- testing commands
  1. pytest test/distributed/test_c10d.py -k test_sparse_gradients -vs
  2.  pytest test/distributed/test_c10d.py -k test_mixed_dense_sparse_gradients -vs
  3.  pytest test/distributed/test_c10d.py -k test_ddp_comm_hook_sparse_gradients -vs
  4.  pytest test/distributed/test_c10d.py -k test_ddp_comm_hook_mixed_dense_sparse_gradients -vs

[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

work for issue https://github.com/pytorch/pytorch/issues/54445

- Adds a MixedDenseSparseGradientModule 
- refactored SparseGradientModule tests
- added MixedDenseSparseGradientModule tests
- testing commands
  1. pytest test/distributed/test_c10d.py -k test_sparse_gradients -vs
  2.  pytest test/distributed/test_c10d.py -k test_mixed_dense_sparse_gradients -vs
  3.  pytest test/distributed/test_c10d.py -k test_ddp_comm_hook_sparse_gradients -vs
  4.  pytest test/distributed/test_c10d.py -k test_ddp_comm_hook_mixed_dense_sparse_gradients -vs

[ghstack-poisoned]

gcramer23 committed:
Update on "MixedDenseSparseGradientModule"

work for issue https://github.com/pytorch/pytorch/issues/54445

- Adds a MixedDenseSparseGradientModule 
- refactored SparseGradientModule tests
- added MixedDenseSparseGradientModule tests
- testing commands
  1. pytest test/distributed/test_c10d.py -k test_sparse_gradients -vs
  2.  pytest test/distributed/test_c10d.py -k test_mixed_dense_sparse_gradients -vs
  3.  pytest test/distributed/test_c10d.py -k test_ddp_comm_hook_sparse_gradients -vs
  4.  pytest test/distributed/test_c10d.py -k test_ddp_comm_hook_mixed_dense_sparse_gradients -vs

[ghstack-poisoned]

mrshenli reviewed (approved):
LGTM! And thanks for cleanup duplicated code with existing tests.

mrshenli commented:
Could you please import the PR as a diff? I will stamp the diff too.

wayi1 reviewed (approved):
Thanks for the refactoring!
